metrics/metrics.go

```diff
@@ -38,6 +40,33 @@ func NewRegistry(version string) *prometheus.Registry {
 	return registry
 }
 
+var providerMetricsKey = "provider-metrics"
+
+func ProviderCallMetricsFromContext(c context.Context) *prometheus.HistogramVec {
+	if raw := c.Value(providerMetricsKey); raw != nil {
+		return raw.(*prometheus.HistogramVec) // nolint:forcetypeassert
+	}
+	return nil
+}
+
+type ProviderActionOpts struct {
+	ProviderID      uid.ID
+	OrgID           uid.ID
+	Action          string
+	DurationSeconds float64
+}
+
+func ProviderAction(c context.Context, opts ProviderActionOpts) {
+	m := ProviderCallMetricsFromContext(c)
+	if m != nil {
+		m.With(prometheus.Labels{
+			"provider_id":     opts.ProviderID.String(),
+			"organization_id": opts.OrgID.String(),
```

Comment:
I believe these identifiers are going to be too high cardinality for metrics. They're fine for logs, but for metrics we probably need to settle for the provider kind. In many cases the provider kind should be sufficient I guess, because all the providers of the same kind are probably hitting the same servers.